quint/mempool/Consensus.qnt

```diff
@@ -0,0 +1,155 @@
+// -*- mode: Bluespec; -*-
+// The common consensus abstraction where processes agree on some blocks (i.e., set of transactions).
+
+module Consensus{
+
+    import Spells.* from "Spells"
+    import System.* from "System"
+
+    type ConsensusState = {
+      proposed: Process -> Set[Tx],
+      decided: Process -> Set[Tx]
+    }
+
+    pure def newConsensus(P: Set[Process]): ConsensusState = {
+      proposed: P.mapBy(p => Set()),
+      decided: P.mapBy(p => Set())
+    }
+
+    pure def processes(state: ConsensusState): Set[Process] = state.proposed.keys()
+
+    pure def hasProposed(state: ConsensusState, p: Process): bool = state.proposed.get(p).size() > 0
+    
+    pure def hasDecided(state: ConsensusState, p: Process): bool = state.decided.get(p).size() > 0
+
+    pure def isProposed(state: ConsensusState, t: Tx): bool = state.proposed.keys().filter(p => state.proposed.get(p).contains(t)).size() > 0
+
+    pure def isDecided(state: ConsensusState, t: Tx): bool = state.decided.keys().filter(p => state.decided.get(p).contains(t)).size() > 0
+
+    pure def hasDecision(state: ConsensusState): bool = state.decided.keys().filter(p => state.decided.get(p).size()>0).size() > 0
+
+    pure def decisionOf(state: ConsensusState, p: Process): Set[Tx] = state.decided.get(p)
+
+    pure def proposalOf(state: ConsensusState, p: Process): Set[Tx] = state.proposed.get(p)
+
+    pure def getProposed(state: ConsensusState): Set[Tx] = flatten(mapValues(state.proposed))
+
+    pure def getDecided(state: ConsensusState): Set[Tx] = flatten(mapValues(state.decided))
+
+    pure def mayPropose(state: ConsensusState, p: Process, txs: Set[Tx]): bool = processes(state).contains(p) and not(hasProposed(state, p)) and txs.size()!=0
+
+    pure def mayDecide(state: ConsensusState, p: Process): bool = processes(state).contains(p) and state.hasProposed(p) and not(state.hasDecided(p))
+
+    pure def propose(state: ConsensusState, p: Process, txs: Set[Tx]): ConsensusState = {
+      val newProposed = state.proposed.set(p, txs)
+      state.with("proposed", newProposed)
```

Comment:
There is a new syntax for record manipulation that is more pleasing (at least to me).
This would become `{...state, proposed:newProposed}`

Reply:
I guess I just dislike the quotes :)

Reply:
I changed the syntax as it is indeed strange the use of quotes.